File: src/main/java/me/sd_master92/customvoting/subjects/VoteParty.java
package me.sd_master92.customvoting.subjects;

import me.sd_master92.customvoting.Main;
import me.sd_master92.customvoting.constants.Data;
import me.sd_master92.customvoting.constants.Messages;
import me.sd_master92.customvoting.constants.Settings;
import me.sd_master92.customvoting.constants.enumerations.SoundType;
import me.sd_master92.customvoting.constants.enumerations.VotePartyType;
import me.sd_master92.customvoting.gui.GUI;
import org.bukkit.*;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.*;

public class VoteParty
{
    public static final ItemStack VOTE_PARTY_ITEM = GUI.createItem(Material.ENDER_CHEST,
            ChatColor.LIGHT_PURPLE +
                    "Vote Party Chest",
            ChatColor.GRAY + "Place this chest somewhere in the sky.;" + ChatColor.GRAY + "The contents of this chest" +
                    " will;" +
                    ChatColor.GRAY + "start dropping when the voteparty starts.");
    private static final List<VoteParty> queue = new ArrayList<>();
    private static boolean isActive = false;
    private final Main plugin;
    private final int votePartyType;
    private int count;

    public VoteParty(Main plugin)
    {
        this.plugin = plugin;
        votePartyType = plugin.getConfig().getNumber(Settings.VOTE_PARTY_TYPE);
        count = plugin.getConfig().getNumber(Settings.VOTE_PARTY_COUNTDOWN);
    }

    public void start()
    {
        queue.add(this);
        if (!isActive)
        {
            isActive = true;
            new BukkitRunnable()
            {
                @Override
                public void run()
                {
                    switch (count)
                    {
                        case 30:
                        case 15:
                        case 10:
                            Map<String, String> placeholders = new HashMap<>();
                            placeholders.put("%TIME%", "" + count);
                            SoundType.NOTIFY.playForAll(plugin);
                            plugin.getServer().broadcastMessage(Messages.VOTE_PARTY_COUNTDOWN.getMessage(plugin, placeholders));
                            break;
                        case 5:
                        case 4:
                        case 3:
                        case 2:
                        case 1:
                            placeholders = new HashMap<>();
                            placeholders.put("%TIME%", "" + count);
                            placeholders.put("%s%", count == 1 ? "" : "s");
                            SoundType.NOTIFY.playForAll(plugin);
                            plugin.getServer().broadcastMessage(Messages.VOTE_PARTY_COUNTDOWN_ENDING.getMessage(plugin, placeholders));
                            break;
                        case 0:
                            SoundType.VOTE_PARTY_START.playForAll(plugin);
                            plugin.getServer().broadcastMessage(Messages.VOTE_PARTY_START.getMessage(plugin));
                            if (votePartyType == VotePartyType.RANDOM_CHEST_AT_A_TIME.getValue())
                            {
                                dropChestsRandomly();
                            } else if (votePartyType == VotePartyType.ADD_TO_INVENTORY.getValue())
                            {
                                addToInventory();
                            } else
                            {
                                dropChests();
                            }
                            cancel();
                    }
                    count--;
                }
            }.runTaskTimer(plugin, 0, 20);
        }
    }

    private void stop()
    {
        isActive = false;
        queue.remove(0);
        if (queue.size() > 0)
        {
            queue.get(0).start();
        }
    }

    private void dropChests()
    {

        Map<String, Location> locations = plugin.getData().getLocations(Data.VOTE_PARTY);

        Random random = new Random();
        Map<Integer, Boolean> tasks = new HashMap<>();

        for (String key : locations.keySet())
        {
            List<ItemStack> chest =
                    new ArrayList<>(Arrays.asList(plugin.getData().getItems(Data.VOTE_PARTY + "." + key)));

            Location loc = locations.get(key).clone().add(0.5, 0, 0.5);
            Location dropLoc = new Location(loc.getWorld(), loc.getX(), loc.getY() - 1, loc.getZ());
            Location fireworkLoc = new Location(loc.getWorld(), loc.getX(), loc.getY() + 1, loc.getZ());

            new BukkitRunnable()
            {
                @Override
                public void run()
                {
                    if (!tasks.containsKey(getTaskId()))
                    {
                        tasks.put(getTaskId(), false);
                    }
                    if (votePartyType == VotePartyType.ALL_CHESTS_AT_ONCE.getValue() || !tasks.containsValue(true) || tasks.get(getTaskId()))
                    {
                        if (votePartyType == VotePartyType.ONE_CHEST_AT_A_TIME.getValue())
                        {
                            tasks.put(getTaskId(), true);
                        }
                        if (chest.size() > 0)
                        {
                            int i = random.nextInt(chest.size());
                            if (dropLoc.getWorld() != null)
                            {
                                dropLoc.getWorld().dropItem(dropLoc, chest.remove(i));
                            }
                            if (random.nextInt(2) == 0)
                            {
                                CustomVote.shootFirework(plugin, fireworkLoc);
                            }
                        } else
                        {
                            tasks.remove(getTaskId());
                            if (tasks.isEmpty())
                            {
                                plugin.getServer().broadcastMessage(Messages.VOTE_PARTY_END.getMessage(plugin));
                                stop();
                            }
                            cancel();
                        }
                    }
                }
            }.runTaskTimer(plugin, 0, 10);
        }
    }

    private void dropChestsRandomly()
    {
        Map<String, Location> locations = plugin.getData().getLocations(Data.VOTE_PARTY);
        Map<String, List<ItemStack>> chests = new HashMap<>();
        List<String> keys = new ArrayList<>(locations.keySet());
        keys.forEach(key ->
        {
            ItemStack[] items = plugin.getData().getItems(
                    Data.VOTE_PARTY + "." + key);
            if (items.length > 0)
            {
                chests.put(key,
                        new ArrayList<>(Arrays.asList(items)));
            }
        });

        Random random = new Random();
        new BukkitRunnable()
        {
            @Override
            public void run()
            {
                if (!keys.isEmpty() && !chests.isEmpty())
                {
                    String key = keys.get(random.nextInt(keys.size()));
                    List<ItemStack> chest = chests.get(key);
                    Location loc = locations.get(key).clone().add(0.5, 0, 0.5);
                    Location dropLoc = new Location(loc.getWorld(), loc.getX(), loc.getY() - 1, loc.getZ());
                    Location fireworkLoc = new Location(loc.getWorld(), loc.getX(), loc.getY() + 1, loc.getZ());

                    if (dropLoc.getWorld() != null)
                    {
                        dropLoc.getWorld().dropItem(dropLoc, chest.remove(random.nextInt(chest.size())));
                        if (chest.isEmpty())
                        {
                            keys.remove(key);
                        } else
                        {
                            chests.put(key, chest);
                        }
                    }
                    if (random.nextInt(3) == 0)
                    {
                        CustomVote.shootFirework(plugin, fireworkLoc);
                    }
                } else
                {
                    plugin.getServer().broadcastMessage(Messages.VOTE_PARTY_END.getMessage(plugin));
                    stop();
                    cancel();
                }
            }
        }.runTaskTimer(plugin, 0, 10);
    }

    private void addToInventory()
    {
        Random random = new Random();
        Map<Integer, Boolean> tasks = new HashMap<>();

        for (String key : plugin.getData().getLocations(Data.VOTE_PARTY).keySet())
        {
            List<ItemStack> chest =
                    new ArrayList<>(Arrays.asList(plugin.getData().getItems(Data.VOTE_PARTY + "." + key)));

            new BukkitRunnable()
            {
                @Override
                public void run()
                {
                    if (!tasks.containsKey(getTaskId()))
                    {
                        tasks.put(getTaskId(), false);
                    }
                    if (!tasks.containsValue(true) || tasks.get(getTaskId()))
                    {
                        tasks.put(getTaskId(), true);
                        if (chest.size() > 0)
                        {
                            List<Player> players = new ArrayList<>(Bukkit.getOnlinePlayers());
                            Player player = players.get(random.nextInt(players.size()));
                            ItemStack item = chest.remove(random.nextInt(chest.size()));
                            for(ItemStack left : player.getInventory().addItem(item).values())
                            {
                                player.getWorld().dropItem(player.getLocation(), left);
                            }
                            SoundType.PICKUP.play(plugin, player);
                        } else
                        {
                            tasks.remove(getTaskId());
                            if (tasks.isEmpty())
                            {
                                plugin.getServer().broadcastMessage(Messages.VOTE_PARTY_END.getMessage(plugin));
                                stop();
                            }
                            cancel();
                        }
                    }
                }
            }.runTaskTimer(plugin, 0, 10);
        }
    }
}

File: src/main/kotlin/me/sd_master92/customvoting/subjects/VoteParty.kt
package me.sd_master92.customvoting.subjects

import me.sd_master92.customvoting.CV
import me.sd_master92.customvoting.constants.Data
import me.sd_master92.customvoting.constants.Messages
import me.sd_master92.customvoting.constants.Settings
import me.sd_master92.customvoting.constants.enumerations.SoundType
import me.sd_master92.customvoting.constants.enumerations.VotePartyType
import me.sd_master92.customvoting.gui.GUI
import me.sd_master92.customvoting.helpers.ParticleHelper
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.scheduler.BukkitRunnable
import java.util.*
import java.util.function.Consumer

class VoteParty(private val plugin: CV)
{
    private val votePartyType: Int = plugin.config.getNumber(Settings.VOTE_PARTY_TYPE)
    private var count: Int = plugin.config.getNumber(Settings.VOTE_PARTY_COUNTDOWN)
    fun start()
    {
        queue.add(this)
        if (!isActive)
        {
            isActive = true
            object : BukkitRunnable()
            {
                override fun run()
                {
                    when (count)
                    {
                        30, 15, 10    ->
                        {
                            if (!plugin.config.getBoolean(Settings.DISABLED_BROADCAST_VOTE_PARTY_COUNTDOWN))
                            {
                                val placeholders = HashMap<String, String>()
                                placeholders["%TIME%"] = "" + count
                                SoundType.NOTIFY.playForAll(plugin)
                                plugin.server.broadcastMessage(
                                    Messages.VOTE_PARTY_COUNTDOWN.getMessage(
                                        plugin,
                                        placeholders
                                    )
                                )
                            }
                        }
                        5, 4, 3, 2, 1 ->
                        {
                            if (!plugin.config.getBoolean(Settings.DISABLED_BROADCAST_VOTE_PARTY_COUNTDOWN_ENDING))
                            {
                                val placeholders = HashMap<String, String>()
                                placeholders["%TIME%"] = "" + count
                                placeholders["%s%"] = if (count == 1) "" else "s"
                                SoundType.NOTIFY.playForAll(plugin)
                                plugin.server.broadcastMessage(
                                    Messages.VOTE_PARTY_COUNTDOWN_ENDING.getMessage(
                                        plugin,
                                        placeholders
                                    )
                                )
                            }
                        }
                        0             ->
                        {
                            SoundType.VOTE_PARTY_START.playForAll(plugin)
                            plugin.server.broadcastMessage(Messages.VOTE_PARTY_START.getMessage(plugin))
                            when (votePartyType)
                            {
                                VotePartyType.RANDOM_CHEST_AT_A_TIME.value ->
                                {
                                    dropChestsRandomly()
                                }
                                VotePartyType.ADD_TO_INVENTORY.value       ->
                                {
                                    addToInventory()
                                }
                                else                                       ->
                                {
                                    dropChests()
                                }
                            }
                            cancel()
                        }
                    }
                    count--
                }
            }.runTaskTimer(plugin, 0, 20)
        }
    }

    private fun stop()
    {
        isActive = false
        queue.removeFirstOrNull()
        if (queue.isNotEmpty())
        {
            queue[0].start()
        }
    }

    private fun dropChests()
    {
        val locations = plugin.data.getLocations(Data.VOTE_PARTY)
        val random = Random()
        val tasks: MutableMap<Int, Boolean> = HashMap()
        for (key in locations.keys)
        {
            val chest: MutableList<ItemStack> = ArrayList(listOf(*plugin.data.getItems(Data.VOTE_PARTY + "." + key)))
            val loc = locations[key]!!.clone().add(0.5, 0.0, 0.5)
            val dropLoc = Location(loc.world, loc.x, loc.y - 1, loc.z)
            val fireworkLoc = Location(loc.world, loc.x, loc.y + 1, loc.z)
            object : BukkitRunnable()
            {
                override fun run()
                {
                    if (!tasks.containsKey(taskId))
                    {
                        tasks[taskId] = false
                    }
                    if (votePartyType == VotePartyType.ALL_CHESTS_AT_ONCE.value || !tasks.containsValue(true) || tasks[taskId]!!)
                    {
                        if (votePartyType == VotePartyType.ONE_CHEST_AT_A_TIME.value)
                        {
                            tasks[taskId] = true
                        }
                        if (chest.isNotEmpty())
                        {
                            val i = random.nextInt(chest.size)
                            if (dropLoc.world != null)
                            {
                                dropLoc.world!!.dropItem(dropLoc, chest.removeAt(i))
                            }
                            if (random.nextInt(2) == 0)
                            {
                                ParticleHelper.shootFirework(plugin, fireworkLoc)
                            }
                        } else
                        {
                            tasks.remove(taskId)
                            if (tasks.isEmpty())
                            {
                                plugin.server.broadcastMessage(Messages.VOTE_PARTY_END.getMessage(plugin))
                                stop()
                            }
                            cancel()
                        }
                    }
                }
            }.runTaskTimer(plugin, 0, 10)
        }
    }

    private fun dropChestsRandomly()
    {
        val locations = plugin.data.getLocations(Data.VOTE_PARTY)
        val chests: MutableMap<String, MutableList<ItemStack>> = HashMap()
        val keys: MutableList<String> = ArrayList(locations.keys)
        keys.forEach(Consumer { key: String ->
            val items = plugin.data.getItems(
                Data.VOTE_PARTY + "." + key
            )
            if (items.isNotEmpty())
            {
                chests[key] = ArrayList(listOf(*items))
            }
        })
        val random = Random()
        object : BukkitRunnable()
        {
            override fun run()
            {
                if (keys.isNotEmpty() && chests.isNotEmpty())
                {
                    val key = keys[random.nextInt(keys.size)]
                    val chest = chests[key]!!
                    val loc = locations[key]!!.clone().add(0.5, 0.0, 0.5)
                    val dropLoc = Location(loc.world, loc.x, loc.y - 1, loc.z)
                    val fireworkLoc = Location(loc.world, loc.x, loc.y + 1, loc.z)
                    if (dropLoc.world != null)
                    {
                        dropLoc.world!!.dropItem(dropLoc, chest.removeAt(random.nextInt(chest.size)))
                        if (chest.isEmpty())
                        {
                            keys.remove(key)
                        } else
                        {
                            chests[key] = chest
                        }
                    }
                    if (random.nextInt(3) == 0)
                    {
                        ParticleHelper.shootFirework(plugin, fireworkLoc)
                    }
                } else
                {
                    plugin.server.broadcastMessage(Messages.VOTE_PARTY_END.getMessage(plugin))
                    stop()
                    cancel()
                }
            }
        }.runTaskTimer(plugin, 0, 10)
    }

    private fun addToInventory()
    {
        val random = Random()
        val tasks: MutableMap<Int, Boolean> = HashMap()
        for (key in plugin.data.getLocations(Data.VOTE_PARTY).keys)
        {
            val chest: MutableList<ItemStack> = ArrayList(listOf(*plugin.data.getItems(Data.VOTE_PARTY + "." + key)))
            object : BukkitRunnable()
            {
                override fun run()
                {
                    if (!tasks.containsKey(taskId))
                    {
                        tasks[taskId] = false
                    }
                    if (!tasks.containsValue(true) || tasks[taskId]!!)
                    {
                        tasks[taskId] = true
                        if (chest.isNotEmpty())
                        {
                            val players: List<Player> = ArrayList(Bukkit.getOnlinePlayers())
                            val player = players[random.nextInt(players.size)]
                            val item: ItemStack = chest.removeAt(random.nextInt(chest.size))
                            for (left in player.inventory.addItem(item).values)
                            {
                                player.world.dropItem(player.location, left!!)
                            }
                            SoundType.PICKUP.play(plugin, player)
                        } else
                        {
                            tasks.remove(taskId)
                            if (tasks.isEmpty())
                            {
                                plugin.server.broadcastMessage(Messages.VOTE_PARTY_END.getMessage(plugin))
                                stop()
                            }
                            cancel()
                        }
                    }
                }
            }.runTaskTimer(plugin, 0, 10)
        }
    }

    companion object
    {
        val VOTE_PARTY_ITEM = GUI.createItem(
            Material.ENDER_CHEST, ChatColor.LIGHT_PURPLE.toString() +
                    "Vote Party Chest",
            ChatColor.GRAY.toString() + "Place this chest somewhere in the sky.;" + ChatColor.GRAY + "The contents of this chest" +
                    " will;" +
                    ChatColor.GRAY + "start dropping when the voteparty starts."
        )
        private val queue: MutableList<VoteParty> = ArrayList()
        private var isActive = false
    }

}